### DISCUSSION TITLE
Infinite NNodes() rewards being returned when trying to initialise a new environment from the current instance state

### DISCUSSION THREAD

**Discussion by cwfparsonson** [ds4dm/ml4co-competition#43] [Q&A] [answered]
Hi,

I would like to use a branching agent to solve a set cover instance in a `base_env` branching environment, but at each step in the `base_env` environment I would like to 'rollout' the agent and solve the problem from the current step in a separate `rollout_env` environment. To do this, I am getting the `pyscipopt.Model` from the `ecole` `base_env` environment, then initialising a new rollout model `rollout_m` from this `sourceModel`, converting `rollout_m` to an `ecole.scip.Model()` object, and then resetting a new `rollout_env` from this `rollout_m` ecole model. However, this rollout approach seems to change the original instance in `base_env`, because even though I thought I am make a new separate copy of the instance, `base_env` returns infinite or 0 agent rewards (presumably because the underlying instance in `base_env` is somehow being solved in `rollout_env`): 

```python
instances = ecole.instance.SetCoverGenerator(n_rows=100, n_cols=100, density=0.05)
base_env = EcoleBranching()
base_env.seed(0)
agent = PseudocostBranchingAgent()
while True:
    # find an instance not pre-solved by environment
    print('\n >>> New Instance <<<')
    obs, counter = None, 0
    while obs is None:
        counter += 1
        print(counter)
        instance = next(instances)
        instance_before_reset = instance.copy_orig()
        obs, action_set, reward, done, info = base_env.reset(instance)
        
    # use an agent to solve the instance in base_env
    while not done:

        # rollout agent from current step to episode termination
        print('>> New Rollout <<')
        curr_state = base_env.model.as_pyscipopt()
        rollout_m = pyscipopt.Model(sourceModel=curr_state, globalcopy=True)
        rollout_env = EcoleBranching()
        rollout_env.seed(0)
        rollout_m = ecole.scip.Model.from_pyscipopt(rollout_m)
        rollout_obs, rollout_action_set, rollout_reward, rollout_done, rollout_info = rollout_env.reset(rollout_m)
        while not rollout_done:
            rollout_action = agent.action_select(rollout_action_set, rollout_env.model, rollout_done)
            rollout_action = rollout_action_set[rollout_action]
            rollout_obs, rollout_action_set, rollout_reward, rollout_done, rollout_info = rollout_env.step(rollout_action)
            print('rollout reward (num nodes): {}'.format(rollout_reward['num_nodes']))

        # get agent action at current step
        action = agent.action_select(action_set, base_env.model, done)
        action = action_set[action]
            
        # take step in agent environment
        obs, action_set, reward, done, info = base_env.step(action)
        print('agent reward (num nodes): {}'.format(reward['num_nodes']))
        
        if reward['num_nodes'] < -1e10:
            raise Exception('Agent reward {} invalid'.format(reward['num_nodes']))
```

Output: 
```
 >>> New Instance <<<
1
2
>> New Rollout <<
rollout reward (num nodes): -2.0
rollout reward (num nodes): -2.0
agent reward (num nodes): -1.8446744073709552e+19
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-15-caadaf9b8a43> in <module>
     96 
     97         if reward['num_nodes'] < -1e10:
---> 98             raise Exception('Agent reward {} invalid'.format(reward['num_nodes']))

Exception: Agent reward -1.8446744073709552e+19 invalid
```

If I stop doing rollouts by removing the block of code below the comment 'rollout agent from current step to epsiode termination', then I no longer get these infinite `NNodes()` agent rewards and the code runs fine:

```python
instances = ecole.instance.SetCoverGenerator(n_rows=100, n_cols=100, density=0.05)
base_env = EcoleBranching()
base_env.seed(0)
agent = PseudocostBranchingAgent()
while True:
    # find an instance not pre-solved by environment
    print('\n >>> New Instance <<<')
    obs, counter = None, 0
    while obs is None:
        counter += 1
        print(counter)
        instance = next(instances)
        instance_before_reset = instance.copy_orig()
        obs, action_set, reward, done, info = base_env.reset(instance)
        
    # use an agent to solve the instance in base_env
    while not done:
        # get agent action at current step
        action = agent.action_select(action_set, base_env.model, done)
        action = action_set[action]
            
        # take step in agent environment
        obs, action_set, reward, done, info = base_env.step(action)
        print('agent reward (num nodes): {}'.format(reward['num_nodes']))
        
        if reward['num_nodes'] < -1e10:
            raise Exception('Agent reward {} invalid'.format(reward['num_nodes']))
```

Why is ecole returning infinite NNodes() when I use the above rollout implementation? Is there a better way to initialise new ecole environments from a given partially solved instance at a certain step in an ecole environment?

For completeness, here are the `PseudocostBranchingAgent` and `EcoleBranching` classes used above:

```python
import ecole
import pyscipopt

# ecole branching environment
DEFAULT_SCIP_PARAMS = {'separating/maxrounds': 0,
                       'separating/maxroundsroot': 0,
                       'separating/maxcuts': 0,
                       'separating/maxcutsroot': 0,
                       'presolving/maxrounds': 0,
                       'presolving/maxrestarts': 0,
                       'propagating/maxrounds':0,
                       'propagating/maxroundsroot':0,
                       'lp/initalgorithm':'d',
                       'lp/resolvealgorithm':'d',
                       'limits/time': 3600}

class EcoleBranching(ecole.environment.Branching):
    def __init__(self,
                 observation_function=(
                     ecole.observation.NodeBipartite()
                 ),
                 information_function=({
                     'lp_iterations': ecole.reward.LpIterations().cumsum(),
                     'num_nodes': ecole.reward.NNodes().cumsum(),
                     'solving_time': ecole.reward.SolvingTime().cumsum(),
                     'dual_integral': ecole.reward.DualIntegral(),
                     'primal_dual_integral': ecole.reward.PrimalDualIntegral(),
                 }),
                 reward_function=({
                     'num_nodes': -ecole.reward.NNodes(),
                     'lp_iterations': -ecole.reward.LpIterations(),
                     'primal_integral': -ecole.reward.PrimalIntegral(),
                     'dual_integral': -ecole.reward.DualIntegral(),
                     'primal_dual_integral': -ecole.reward.PrimalDualIntegral(),
                 }),
                 scip_params=DEFAULT_SCIP_PARAMS):
        super(EcoleBranching, self).__init__(observation_function=observation_function,
                                             information_function=information_function,
                                             reward_function=reward_function,
                                             scip_params=scip_params)

# branching agent
class PseudocostBranchingAgent:
    def __init__(self, name='pc'):
        self.name = name
        self.pc_branching_function = ecole.observation.Pseudocosts()

    def extract(self, model, done):
        return self.pc_branching_function.extract(model, done)

    def action_select(self, action_set, model, done):
        scores = self.extract(model, done)
        action = scores[action_set].argmax()
        return action
```

**Comment by AntoinePrv** [comment, marked as answer]
Hi @cwfparsonson,

My best guess is that you have mutable default arguments in `EcoleBranching.__init__`.
In your `EcoleBranching`, the default `reward_function` is being shared among all your environment instances (that's how Python works), and because they are mutable... 💣 💥 
You can read more about this for instance [in this post](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments).

Instead, you should do something like
```python
class EcoleBranching(ecole.environment.Branching):
    def __init__(
        self,
        observation_function="default",
        information_function="default"
        reward_function="default",
        scip_params="default",
    ):
         if reward_function == "default":
             reward_function = reward_function=({
                     'num_nodes': -ecole.reward.NNodes(),
                     'lp_iterations': -ecole.reward.LpIterations(),
                     'primal_integral': -ecole.reward.PrimalIntegral(),
                     'dual_integral': -ecole.reward.DualIntegral(),
                     'primal_dual_integral': -ecole.reward.PrimalDualIntegral(),
                 })
         ...
        super(EcoleBranching, self).__init__(
            observation_function=observation_function,
            information_function=information_function,
            reward_function=reward_function,
            scip_params=scip_params,
        )
```

**Reply by cwfparsonson** [reply]
> Hi @cwfparsonson,
> 
> My best guess is that you have mutable default arguments in `EcoleBranching.__init__`.
> In your `EcoleBranching`, the default `reward_function` is being shared among all your environment instances (that's how Python works), and because they are mutable...  
> You can read more about this for instance [in this post](https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments).
> 
> Instead, you should do something like
> 
> ```python
> class EcoleBranching(ecole.environment.Branching):
>     def __init__(
>         self,
>         observation_function="default",
>         information_function="default"
>         reward_function="default",
>         scip_params="default",
>     ):
>          if reward_function == "default":
>              reward_function = reward_function=({
>                      'num_nodes': -ecole.reward.NNodes(),
>                      'lp_iterations': -ecole.reward.LpIterations(),
>                      'primal_integral': -ecole.reward.PrimalIntegral(),
>                      'dual_integral': -ecole.reward.DualIntegral(),
>                      'primal_dual_integral': -ecole.reward.PrimalDualIntegral(),
>                  })
>          ...
>         super(EcoleBranching, self).__init__(
>             observation_function=observation_function,
>             information_function=information_function,
>             reward_function=reward_function,
>             scip_params=scip_params,
>         )
> ```

Thank you @AntoinePrv, I did not realise this functionality in python! Your solution works perfectly.

One question I have though is that it seems with my rollout implementation SCIP discards the B&B tree state and initialises a new one from scratch (i.e. the number of nodes in the search tree returns to 1 (root node) when initialising a new environment from the current SCIP state, but the root node has the same primal-dual bound as the partially solved instance so not all progress is 'lost'). Is there a way to initialise a new environment instance from a partially solved one without discarding the B&B tree state (e.g. do something equivalent to `rollout_env = copy.deepcopy(base_env)`)? Ideally I would like the `rollout_env` and the underlying partially solved SCIP instance to be the exact same as the current `base_env` when I initialise a new `rollout_env` at each `base_env` step.

**Reply by AntoinePrv** [reply]
This is a topic we are interested in but haven't investigated.
It boils down to find a way to copy the true SCIP state and at the moment I'm not sure that's possible with SCIP itself.
(`copy.deepcopy(base_env)` is not enabled for environments because we haven't found that function to truly copy the SCIP data).

I believe the closest thing to doing rollouts in SCIP is [`SCIPstartProbing`](https://www.scipopt.org/doc-7.0.3/html/group__PublicProbingMethods.php#ga91d6fa45519fc768d1e8460ba8015157) but it is not as general, and does not leave the state truly unchanged.

**Reply by cwfparsonson** [reply]
> This is a topic we are interested in but haven't investigated.
> It boils down to find a way to copy the true SCIP state and at the moment I'm not sure that's possible with SCIP itself.
> (`copy.deepcopy(base_env)` is not enabled for environments because we haven't found that function to truly copy the SCIP data).
> 
> I believe the closest thing to doing rollouts in SCIP is [`SCIPstartProbing`](https://www.scipopt.org/doc-7.0.3/html/group__PublicProbingMethods.php#ga91d6fa45519fc768d1e8460ba8015157) but it is not as general, and does not leave the state truly unchanged.

Okay thank you IL will have a look at `SCIPstartProbing`
